x/merkledb/codec.go

```diff
@@ -222,204 +266,125 @@ func decodeDBNode(b []byte, n *dbNode) error {
 			hasValue:      hasValue,
 		}
 	}
-	if src.Len() != 0 {
+	if len(r.b) != 0 {
 		return errExtraSpace
 	}
 	return nil
 }
 
-func encodeBool(dst *bytes.Buffer, value bool) {
-	bytesValue := falseBytes
-	if value {
-		bytesValue = trueBytes
+func decodeKey(b []byte) (Key, error) {
+	r := codecReader{
+		b:    b,
+		copy: true,
 	}
-	_, _ = dst.Write(bytesValue)
-}
-
-func decodeBool(src *bytes.Reader) (bool, error) {
-	boolByte, err := src.ReadByte()
-	switch {
-	case err == io.EOF:
-		return false, io.ErrUnexpectedEOF
-	case err != nil:
-		return false, err
-	case boolByte == trueByte:
-		return true, nil
-	case boolByte == falseByte:
-		return false, nil
-	default:
-		return false, errInvalidBool
+	key, err := r.Key()
+	if err != nil {
+		return Key{}, err
 	}
+	if len(r.b) != 0 {
+		return Key{}, errExtraSpace
+	}
+	return key, nil
 }
 
-func encodeUint(dst *bytes.Buffer, value uint64) {
-	var buf [binary.MaxVarintLen64]byte
-	size := binary.PutUvarint(buf[:], value)
-	_, _ = dst.Write(buf[:size])
+type codecReader struct {
+	b []byte
+	// copy is used to flag to the reader if it is required to copy references
+	// to [b].
+	copy bool
 }
 
-func decodeUint(src *bytes.Reader) (uint64, error) {
-	// To ensure encoding/decoding is canonical, we need to check for leading
-	// zeroes in the varint.
-	// The last byte of the varint we read is the most significant byte.
-	// If it's 0, then it's a leading zero, which is considered invalid in the
-	// canonical encoding.
-	startLen := src.Len()
-	val64, err := binary.ReadUvarint(src)
-	if err != nil {
-		if err == io.EOF {
-			return 0, io.ErrUnexpectedEOF
-		}
-		return 0, err
+func (r *codecReader) Bool() (bool, error) {
+	if len(r.b) < boolLen {
+		return false, io.ErrUnexpectedEOF
 	}
-	endLen := src.Len()
-
-	// Just 0x00 is a valid value so don't check if the varint is 1 byte
-	if startLen-endLen > 1 {
-		if err := src.UnreadByte(); err != nil {
-			return 0, err
-		}
-		lastByte, err := src.ReadByte()
-		if err != nil {
-			return 0, err
-		}
-		if lastByte == 0x00 {
-			return 0, errLeadingZeroes
-		}
+	boolByte := r.b[0]
+	if boolByte > trueByte {
+		return false, errInvalidBool
 	}
 
-	return val64, nil
+	r.b = r.b[boolLen:]
+	return boolByte == trueByte, nil
 }
 
-func encodeMaybeByteSlice(dst *bytes.Buffer, maybeValue maybe.Maybe[[]byte]) {
-	hasValue := maybeValue.HasValue()
-	encodeBool(dst, hasValue)
-	if hasValue {
-		encodeByteSlice(dst, maybeValue.Value())
+func (r *codecReader) Uvarint() (uint64, error) {
+	length, bytesRead := binary.Uvarint(r.b)
+	if bytesRead <= 0 {
+		return 0, io.ErrUnexpectedEOF
 	}
-}
-
-func decodeMaybeByteSlice(src *bytes.Reader) (maybe.Maybe[[]byte], error) {
-	if minMaybeByteSliceLen > src.Len() {
-		return maybe.Nothing[[]byte](), io.ErrUnexpectedEOF
+	if bytesRead > 1 && r.b[bytesRead-1] == 0x00 {
```

Comment:
Probably worth carrying over the comment from the old code about this being a leading zero